### PR TITLE
BDOG-1879 `FrontendAuditFilter` (configurably) audits all headers

### PR DESCRIPTION
**Pull request by tomwadeson**

### CODE REVIEW

File: bootstrap-backend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/backend/filters/BackendAuditFilterSpec.scala [hunk 1]
```diff
@@ -128,23 +132,23 @@ class BackendAuditFilterSpec
...
-        event.tags("X-Request-ID")      shouldBe xRequestId
-        event.tags("X-Session-ID")      shouldBe xSessionId
-        event.tags("Akamai-Reputation") shouldBe akamaiReputation
-        event.detail("deviceID")        shouldBe deviceID
+        event.auditSource                      shouldBe applicationName
+        event.auditType                        shouldBe requestReceived
+        event.tags("X-Request-ID")             shouldBe xRequestId
+        event.tags("X-Session-ID")             shouldBe xSessionId
+        event.tags("Akamai-Reputation")        shouldBe akamaiReputation
+        event.detail.asJsObjectMap("deviceID") shouldBe JsString(deviceID)
```
Comment (colin-lamed) [inline bootstrap-backend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/backend/filters/BackendAuditFilterSpec.scala:151]: Wondering if we can avoid the implicit helper - maybe something like the following
```scala
event.detail.transform((__ \ "deviceID").json.pick[String]) shouldBe deviceID
```
a (non-implicit) helper could be
```scala
event.detail.transform(stringAt("deviceID")) shouldBe deviceID
```

File: bootstrap-frontend-play-28/src/main/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/AuditableRequestHeaders.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    val empty: Redactions =
+      Redactions(Set.empty, Set.empty)
+  }
+
+  def from(headers: Headers, redactions: Redactions): AuditableRequestHeaders = {
+    val updatedCookies =
+      headers
+        .getAll("Cookie")
+        .map(_.split("; ").filterNot { cookie =>
```
Comment (colin-lamed) [inline bootstrap-frontend-play-28/src/main/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/AuditableRequestHeaders.scala:46]: Can we rely on there always being a space following the `;`?
Also can there be spaces around the `=` below too?
(something about whitespace here https://www.rfc-editor.org/rfc/rfc2109#section-4.1)

Looks like the separator could also be a ',' (See https://www.rfc-editor.org/rfc/rfc2109#section-4.3.4)

File: bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/FrontendAuditFilterSpec.scala [hunk 3]
```diff
@@ -579,41 +648,41 @@ class FrontendAuditFilterServerSpec
...
     val event = verifyAndRetrieveEvent
-    event.detail                                 should not be null
-    event.detail.getOrElse(detailKey, "").length should equal(length)
+    event.detail should not be null
+    val valueAsString =
+      event
+        .detail
+        .asJsObjectMap
+        .get(detailKey) match {
+          case Some(JsString(str)) => str
```
Comment (colin-lamed) [inline bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/FrontendAuditFilterSpec.scala:657]: here would benefit from using a pick transformer as mentioned above

File: bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/FrontendAuditFilterSpec.scala [hunk 4]
```diff
@@ -452,34 +513,42 @@ class FrontendAuditFilterSpec
...
 }
 
 class FrontendAuditFilterServerSpec
   extends AnyWordSpec
-     with FrontendAuditFilterInstance
      with Matchers
      with Eventually
      with IntegrationPatience
      with MockitoSugar
      with GuiceOneServerPerTest
```
Comment (colin-lamed) [inline bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/FrontendAuditFilterSpec.scala:528]: are we actually using `GuiceOneServerPerTest`?

File: bootstrap-backend-play-28/src/main/scala/uk/gov/hmrc/play/bootstrap/backend/http/JsonErrorHandler.scala [hunk 5]
```diff
@@ -40,7 +40,7 @@ class JsonErrorHandler @Inject()(
 ) extends HttpErrorHandler
      with BackendHeaderCarrierProvider {
 
-  import httpAuditEvent.dataEvent
+  import httpAuditEvent.extendedEvent
```
Comment (colin-lamed) [inline bootstrap-backend-play-28/src/main/scala/uk/gov/hmrc/play/bootstrap/backend/http/JsonErrorHandler.scala:43]: if we keep the old `dataEvent` too, then we won't break clients. We could however deprecate it.

File: bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/FrontendAuditFilterSpec.scala (hunk 4 above)
Comment (tomwadeson) [reply bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/FrontendAuditFilterSpec.scala:525]: Yes, it's being used in:

```scala
  override def newAppForTest(testData: TestData): Application = {
```

File: bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/BackwardCompatibilitySpec.scala [hunk 6]
```diff
@@ -175,26 +177,28 @@ class BackwardCompatibilitySpec
...
 
     "preserve uk.gov.hmrc.play.bootstrap.filters.frontend.DefaultFrontendAuditFilter" in {
       new uk.gov.hmrc.play.bootstrap.filters.frontend.DefaultFrontendAuditFilter(
-        config            = Configuration(ConfigFactory.load()),
-        controllerConfigs = mock[uk.gov.hmrc.play.bootstrap.config.ControllerConfigs],
-        auditConnector    = mock[uk.gov.hmrc.play.audit.http.connector.AuditConnector],
-        httpAuditEvent    = mock[uk.gov.hmrc.play.bootstrap.config.HttpAuditEvent],
-        mat               = mock[Materializer]
-      )(ec                = mock[ExecutionContext])
+        config                = Configuration(ConfigFactory.load()),
```
Comment (tomwadeson) [inline bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/BackwardCompatibilitySpec.scala:195]: Are we OK with this breaking change given we'll need to bump the version anyway?

File: bootstrap-common/src/test/scala/uk/gov/hmrc/play/bootstrap/config/HttpAuditEventSpec.scala [hunk 7]
```diff
@@ -34,14 +35,14 @@ class HttpAuditEventSpec extends AnyWordSpecLike with Matchers with GuiceOneAppP
     "create a valid audit event with optional headers" in {
       val request =
         FakeRequest().withHeaders("Foo" -> "Bar", "Ehh" -> "Meh", "Surrogate" -> "Cool", "Surrogate" -> "Cool")
-      val event = HttpAuditEventForTest.dataEvent("foo", "bar", request)
-      event.detail.get("surrogate") shouldBe Some("Cool,Cool") //FRIC - play 2.5 now comman delimits multiple headers with the same name into a single header
+      val event = HttpAuditEventForTest.extendedEvent("foo", "bar", request)
```
Comment (colin-lamed) [inline bootstrap-common/src/test/scala/uk/gov/hmrc/play/bootstrap/config/HttpAuditEventSpec.scala:38]: should we keep the `dataEvent` tests too?

File: bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/BackwardCompatibilitySpec.scala (hunk 6 above)
Comment (colin-lamed) [reply bootstrap-frontend-play-28/src/test/scala/uk/gov/hmrc/play/bootstrap/frontend/BackwardCompatibilitySpec.scala:195]: I think it's fine, the Filter is usually injected - the test is just ensuring that references to it in config/injections should still work.

File: bootstrap-common/src/test/scala/uk/gov/hmrc/play/bootstrap/testhelpers/ReadsHelpers.scala [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package uk.gov.hmrc.play.bootstrap.testhelpers
+
+import play.api.libs.json.{Reads, __}
+
+object ReadsHelpers {
+
+  def at[A](field: String)(implicit reads: Reads[A]): Reads[A] =
+    (__ \ field).read(reads)
```
Comment (colin-lamed) [inline bootstrap-common/src/test/scala/uk/gov/hmrc/play/bootstrap/testhelpers/ReadsHelpers.scala:24]: Can we just use `Reads.at`?

File: bootstrap-common/src/test/scala/uk/gov/hmrc/play/bootstrap/testhelpers/ReadsHelpers.scala [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+  def atNullable[A](field: String)(implicit reads: Reads[A]): Reads[Option[A]] =
```
Comment (colin-lamed) [inline bootstrap-common/src/test/scala/uk/gov/hmrc/play/bootstrap/testhelpers/ReadsHelpers.scala:26]: `Reads.nullable`?

File: bootstrap-frontend-play-28/src/main/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/RequestHeaderAuditing.scala [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+      if (updated.nonEmpty)
+        Some(cookieHeaderEncoding.encodeCookieHeader(updated.toSeq))
+      else None
+    }
+
+    val updatedHeaders =
+      updatedCookies
+        .fold(headers)(cookieHeader => headers.remove("Cookie").add("Cookie" -> cookieHeader))
+        .remove(config.redactedHeaders.toSeq: _*)
```
Comment (colin-lamed) [inline bootstrap-frontend-play-28/src/main/scala/uk/gov/hmrc/play/bootstrap/frontend/filters/RequestHeaderAuditing.scala:45]: should we keep the fact that the redacted headers exist, just hide the value?

**Review by colin-lamed — APPROVED** [review]

**Pull request merged**